File: Python Leetcode/287_Find the Duplicate Number.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        slow = nums[0]
        fast = nums[nums[0]]

        while slow != fast:
            slow = nums[slow]
            fast = nums[nums[fast]]

        print("completed")

        slow2 = 0
        while slow2 != slow:
            slow = nums[slow]
            slow2 = nums[slow2]

        return slow

File: Rust Leetcode/287_Find the Duplicate Number.rs
impl Solution {
    pub fn find_duplicate(nums: Vec<i32>) -> i32 {
        let mut slow = nums[0];
        let mut fast = nums[nums[0] as usize];

        while slow != fast {
            slow = nums[slow as usize];
            fast = nums[nums[fast as usize] as usize];
        }

        let mut slow2 = 0;
        while slow2 != slow {
            slow = nums[slow as usize];
            slow2 = nums[slow2 as usize];
        }

        slow
    }
}
